nemo/collections/nlp/modules/common/transformer/transformer_decoders.py

```diff
@@ -0,0 +1,113 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+
+import torch
+import torch.nn as nn
+
+from nemo.collections.common.parts import form_attention_mask
+from nemo.collections.nlp.modules.common.transformer.transformer_modules import MultiHeadAttention, PositionWiseFF
+
+__all__ = ["TransformerDecoder"]
+
+
+class TransformerDecoderBlock(nn.Module):
+    """
+    Building block of Transformer decoder.
+
+    Args:
+        hidden_size: size of the embeddings in the model, also known as d_model
+        inner_size: number of neurons in the intermediate part of feed-forward
+            net, usually is (4-8 x hidden_size) in the papers
+        num_attention_heads: number of heads in multi-head attention
+        attn_score_dropout: probability of dropout applied to attention scores
+        attn_layer_dropout: probability of dropout applied to the output of the
+            attention layers, but before layer normalization
+        ffn_dropout: probability of dropout applied to FFN output
+        hidden_act: activation function used between two linear layers in FFN
+    """
+
+    def __init__(
+        self,
+        hidden_size,
+        inner_size,
+        num_attention_heads=1,
+        attn_score_dropout=0,
+        attn_layer_dropout=0,
+        ffn_dropout=0,
+        hidden_act="relu",
+    ):
+        super().__init__()
+
+        self.first_sub_layer = MultiHeadAttention(
+            hidden_size, num_attention_heads, attn_score_dropout, attn_layer_dropout
+        )
+        self.second_sub_layer = MultiHeadAttention(
+            hidden_size, num_attention_heads, attn_score_dropout, attn_layer_dropout
+        )
+        self.third_sub_layer = PositionWiseFF(hidden_size, inner_size, ffn_dropout, hidden_act)
+
+    def forward(self, decoder_query, decoder_mask, decoder_keys, encoder_states, encoder_mask):
+        self_attn_output = self.first_sub_layer(decoder_query, decoder_keys, decoder_keys, decoder_mask)
+        enc_dec_attn_output = self.second_sub_layer(self_attn_output, encoder_states, encoder_states, encoder_mask)
+        output_states = self.third_sub_layer(enc_dec_attn_output)
+        return output_states
+
+
+class TransformerDecoder(nn.Module):
```

Comment:
Could you please inherit this from NeuralModule instead of nn.Module, define input/output types and add @typecheck() to forward.